packages/queued-request-controller/src/QueuedRequestController.ts

```diff
@@ -263,30 +282,61 @@ export class QueuedRequestController extends BaseController<
    */
   async #processNextBatch() {
     const firstRequest = this.#requestQueue.shift() as QueuedRequest;
-    this.#originOfCurrentBatch = firstRequest.origin;
-    this.#networkClientIdOfCurrentBatch = firstRequest.networkClientId;
-    const batch = [firstRequest.processRequest];
+    this.#originOfCurrentBatch = firstRequest.request.origin;
+    this.#networkClientIdOfCurrentBatch = firstRequest.request.networkClientId;
+    const batch = [firstRequest];
+
+    let networkSwitchError: unknown;
+    try {
+      // If globally selected network is different from origin selected network,
+      // switch network before processing batch
+      await this.#switchNetworkIfNecessary(
+        firstRequest.request.networkClientId,
+      );
+    } catch (error: unknown) {
+      networkSwitchError = error;
+    }
+
+    // If the first request might switch the network, process the request by
+    // itself. If the request does change the network, clear the queue for the
+    // origin since it any remaining requests are now invalidated
+    if (this.#canRequestSwitchNetworkWithoutApproval(firstRequest.request)) {
+      // This hack prevents the next batch from being processed
+      // after this request returns. This is necessary because
+      // we may need to flush the queue before the next set of requests
+      // are batched and processed, which we cannot do without blocking
+      // the queue from continuing by artificially increasing the processing
+      // request count
+      this.#processingRequestCount += 1;
+      firstRequest.processRequest(networkSwitchError);
+      this.#updateQueuedRequestCount();
+      await firstRequest.processedPromise;
+      this.#processingRequestCount -= 1;
```

Comment:
Nit: Maybe worth adding a `try.. finally` here to ensure the request count isn't permanently wrong when a request fails.

I know it's OK now because it can't fail (it can hang, that's it, `reject` is never called though). But it would be one less footgun to leave lying around.

Reply:
added https://github.com/MetaMask/core/pull/4846/commits/11c0a7dc40019250612dcae9e6533a1a74ea46e7